### PR TITLE
Tentative fix for endless-sky/endless-sky#818

### PR DESCRIPTION
**Pull request by db47h**
It works nicely with the various ships I have tested. If you ask your fleet to gather around you, your escorts will get close to your ship then try to match your course and settle after a few seconds. However,  they always very slowly drift away from you (can't match your exact course without cheating), so they'll end up catching up again after a while. This isn't really annoying unless you have hundreds of ships. 

The biggest issue is that when you start flying around, your escorts will have a nasty wriggle... I'm still trying to correct this.

### CODE REVIEW

**Comment by db47h** [comment]
That should be working fine now. With this, the method CircleAround should probably be renamed to FlyAlongWith or something like that... I have not been able to test the part where escorts wandering freely get close to other ships and scan them, but it should be an improvement over the old behavior.

**Comment by jmhorjus** [comment]
I'm going to pull this to one of my branches to test it as soon as I can...very excited about it.

**Comment by jmhorjus** [comment]
The ships, once they have matched your velocity, should continue to match your heading.  That way your whole fleet will turn with you and look awesome, and they'll also be ready for when you decide to accelerate in a new direction.

**Comment by GalacticRuler** [comment]
How well does it work with huge/heavy/slow ships?

**Comment by jmhorjus** [comment]
They try their best to keep up.  If they're far away they just move toward you like before.

**Comment by jmhorjus** [comment]
You need to use the "MoveToward(ship, command, target.Position(), 200, 0.8)" if the ship is far away from it's target instead of your current lines 1030-1035.  
Otherwise a fast moving ship will still zoom right past you instead of smartly turing around ahead of time to get ready for intercept.

**Comment by db47h** [comment]
Slow ships are slower to settle, that's all. I admit that I haven't tried to see what happens withe the slowest possible ship though. 

@jmhorjus feel free to tweak the numbers. All of them are guesstimates. I'll try your suggestion tomorrow with a fleet composed of very fast and very slow ships and see what happens. I'll also see what I can do about turning them in the same direction.

**Comment by db47h** [comment]
Ok, now the escorts match the player's ship heading. Had to tweak a bit the magic values but it works even better. I'll post a video of the result in a few minutes.

@jmhorjus There's no `MoveToward()` anywhere in the codebase I have, so escorts still tend to overshoot the player's ship on the first approach (well, those that can't turn around fast enough)... Care to share? ;-)

**Comment by db47h** [comment]
Video link:

[PR 849 in action](https://youtu.be/Rdq0VlV86vI)

Fleet formations FTW!

**Comment by GalacticRuler** [comment]
Wow...that looks great!. It is a bit chaotic though

**Comment by db47h** [comment]
MoveTo() doesn't work well for moving targets: escorts tend to wriggle back and forth when getting close. Still working on it...

**Comment by jmhorjus** [comment]
I had them MoveTo my position + 120 * my velocity, and it was much better than just moving to my position.

I'm concerned about combat situations though.  The old "mill around in circles" behavior actually had tactical advantages much as I hate to say it.  1.) You could control the position of an escort relative to your ship easily using your own movement, and 2.) moving around helped distribute damage over all of your escorts instead of just one or two taking all the fire from a certain direction and getting disabled.  
Basically just make sure whatever we come up with is tested extensively in combat, not just what looks the best when cruising around.

**Comment by db47h** [comment]
Fully agree with you. For starters, CircleAround() should be kept as is and this new implementation moved to its own method and used only to gather escorts (CircleAround() is used in other situations). And then test. Guess I'll have to go back spending more time playing than hacking the code :)

**Comment by jmhorjus** [comment]
Here's what I think needs to be done:
1.) Improve MoveTo() so that it is able to deal with moving targets. To avoid jitter you need to have "fudge room" between the two behaviors: so, for example, instead of "if(x>5) {do A} else {do B}" you need to have "if(x>6) {do A} else if(x>4) {keep doing whatever you were already doing} else {do B}". This is the kind of logic that prevents the hair-line of a condition from causing a behavior to change every frame. This kind of logic may need to be incorporated elsewhere as well, if jitteriness is encountered. 

2.) Give ships the ability to fly in formation.  They Need to be able to "move to" a location which is an offset of the parent ship's location based on the parent ship's current facing - i.e. "150 units at 180 degree angle" or a simple x,y offset from the parent's location, rotated by the parent's heading.  If the parent ship turns, the following ship will try to get in position relative to the parent's new orientation.  Then "formations" can be algorithms that return these relative locations and, depending on where you are in your parent ship's list of escorts, you get assigned a location to fly in. A formation function needs to return a relative location given any index, since a ship may have any number of escorts.  However, under the hood, it can just rotate through a fixed list of locations - so if you have more than 20 (or whatever number) escorts, then ship 21 will be instructed to fly at the same location as ship 1, etc.

3.) Give the player the ability to cycle between different formations probably with a new hotkey: stay in front, stay behind, keep formation around, circle around me, etc.  There don't need to be a lot of formations, just 2+ formations, one of which can be the existing "circle around".  

@db47h Do you agree? Does this seem doable?

**Comment by db47h** [comment]
Very good thinking about reusing MoveTo() and integrate a target velocity. My only problem is to come up with a good algorithm. Matching the target velocity is easy (just get delta-v, fly in that direction until it cancels out), but doing it while reaching a given point is the tricky part. If someone with a good understanding of the physics involved wants to chip in... 

With the current algorithm, most ships tend to reach the designated target "area" at the requested velocity except bulk freighters that tend to overshoot once. However, it's not good enough if we want to think about implementing 2) and 3), because it can only reach a general area around the player's ship, not a specific point.  Whether that point is moving or not does not really matter.

> To avoid jitter you need to have "fudge room" between the two behaviors: so, for example, instead of "if(x>5) {do A} else {do B}" you need to have "if(x>6) {do A} else if(x>4) {keep doing whatever you were already doing} else {do B}"

I could be wrong, but targeting the target point+velocity should work (at least for the next game step).

Regarding point 2), even if we had a perfectly oiled algorithm, I'm not sure about the result from a game play perspective. Imagine this: you're flying your merry way, ask your ships to adopt a circular defensive formation, then suddenly realize your freighters are lagging behind, so you naturally press down-arrow to turn around and slow down, right? You suddenly have your whole fleet scrambling all over the place to get in their new positions while you turn... I have a feeling it'd be chaotic. Well, may be not too chaotic if your escorts can accelerate enough, that is.

Another problem is the all-or-nothing thrust mechanics. It doesn't allow fine-grained adjustment of the ship's velocity, hence the `if(dV.Length() > .03)` line. In order to compensate for this I was tempted to 1) cheat and set the ship's velocity once reached a close enough velocity (close enough for the player not to notice anything), or 2) add a "partial " thrust command (like it's been done for the heading).

**Comment by jmhorjus** [comment]
My advice: Don't be afraid to cheat a little bit!  I can't believe I never thought of that.  If I'm not mistaken, there's a bit of "cheat" or "tractor beam" or something already going on when you're boarding a ship once you get close enough. In this case even a very small velocity adjustment might really help a lot.  

I'm glad you're continuing to experiment with this. It may seem like a really small and fiddly bit of the code, but it has the potential to effect the game-play experience in a really significant way.  If you're not sure what the best approach is among different options, you might consider doing some plumbing to enable switching between different follow functions during play - it'll not only be a big ask for the game in general (a "switch formation" button), but will probably help a lot with A/B type testing even before it's merged.  
(Normally I'd offer to help with that kind of plumbing job, but I won't be able to give it any serious time until after Easter)

**Comment by endless-sky** [comment]
When you're boarding, your ship uses "magnetic grapples." So, it's not really cheating. :)

The AI doesn't have the ability to command a ship to do anything it's not capable of - all the AI can do is send the same commands that you send by pressing keys on the keyboard.

That said, I like the idea of your escorts [flocking](https://en.wikipedia.org/wiki/Flocking_(behavior)) around your ship, but I think it needs some more tuning. In particular, the real test is whether fast ships can properly keep station around you, not just whether slow ships can match direction with you when you're also moving slowly. And that's where it gets complicated, because a fast ship would need to be able to accelerate to catch up with you, then steer in an S-curve or something to shed velocity to match your speed.

**Comment by db47h** [comment]
Right, next commit I'll try to start with the plumbing to implement formation switching and move the code to MoveTo(). I'd like to get rid of the current "split brain" implementation that does pretty much like the current MoveTo() when far away and "match course" once in range, and I may have an idea on how to do it properly.

Even with the ideal algorithm, as long as ships have to turn around to slow down (anyone really ditching guns just to install front thrusters?), there will be some weird behaviors in formations when turning.

The for those who prefer their escots flocking, or swarming around them, the behavior can be re-implemented with a specific formation algorithm that just moves the target points (why not have them circle around instead of the current back and fro?).

I'm a bit busy right now, but I'm still on it!

**Comment by Wrzlprnft** [comment]
By the way, good job, keep going!

**Comment by Alkallid** [comment]
Ia tgis now obsolete?

**Comment by Alkallid** [comment]
Um, is this. I can't edit.

**Comment by jmhorjus** [comment]
Yes.  MZ has since done his own station-keeping AI function which works very well, and was likely inspired by the experimentation that went on here.  But this pull should be closed now.
